File: web/src/services/mlEngine.ts
import sleepModel from '../assets/models/sleep_model.json';
import phoneModel from '../assets/models/phone_model.json';
import activityModel from '../assets/models/activity_model.json';
import emotionModel from '../assets/models/emotion_model.json';
import masterModel from '../assets/models/master_burnout_model.json';

// Helper for dot product calculation: intercept + sum(coef_i * x_i)
function evaluateLinearModel(coefficients: number[], intercept: number, features: number[]): number {
  let val = intercept;
  for (let i = 0; i < coefficients.length; i++) {
    val += (coefficients[i] || 0) * (features[i] || 0);
  }
  return val;
}

// ---------------------------------------------------------------------
// 1. Sleep Quality & Recovery Predictor
// ---------------------------------------------------------------------
export interface SleepMLPrediction {
  predictedQualityScore: number;
  fatigueIndex: number;
  circadianDisruptionLevel: 'Low' | 'Moderate' | 'High';
}

export function predictSleepQuality(
  bedtimeHour: number,
  wakeTimeHour: number,
  durationHours: number,
  interruptions: number,
  prevScreenTime: number,
  prevLateNight: boolean,
  prevExercise: number,
  prevStress: number
): SleepMLPrediction {
  const inputs = [
    bedtimeHour,
    wakeTimeHour,
    durationHours,
    interruptions,
    prevScreenTime,
    prevLateNight ? 1 : 0,
    prevExercise,
    prevStress,
  ];

  const quality = evaluateLinearModel(
    sleepModel.quality_model.coefficients,
    sleepModel.quality_model.intercept,
    inputs
  );

  const fatigue = evaluateLinearModel(
    sleepModel.fatigue_model.coefficients,
    sleepModel.fatigue_model.intercept,
    inputs
  );

  let disruption: 'Low' | 'Moderate' | 'High' = 'Low';
  if ((bedtimeHour >= 1 && bedtimeHour <= 5) || prevLateNight) {
    disruption = interruptions > 2 ? 'High' : 'Moderate';
  }

  return {
    predictedQualityScore: Math.round(Math.max(10, Math.min(100, quality))),
    fatigueIndex: Math.round(Math.max(0, Math.min(100, fatigue))),
    circadianDisruptionLevel: disruption,
  };
}

// ---------------------------------------------------------------------
// 2. Digital Overuse & Addiction Classifier
// ---------------------------------------------------------------------
export interface DigitalAddictionMLPrediction {
  burnoutContributionScore: number;
  usagePatternClass: 'Healthy' | 'Distracted' | 'Compulsive_Addiction';
  recommendedScreenCapHours: number;
}

export function predictDigitalAddiction(
  totalScreen: number,
  socialMedia: number,
  entertainment: number,
  productive: number,
  pickups: number,
  lateNight: boolean,
  avgSessionMin: number
): DigitalAddictionMLPrediction {
  const inputs = [
    totalScreen,
    socialMedia,
    entertainment,
    productive,
    pickups,
    lateNight ? 1 : 0,
    avgSessionMin,
  ];

  const score = evaluateLinearModel(
    phoneModel.score_model.coefficients,
    phoneModel.score_model.intercept,
    inputs
  );

  const cap = evaluateLinearModel(
    phoneModel.cap_model.coefficients,
    phoneModel.cap_model.intercept,
    inputs
  );

  const finalScore = Math.round(Math.max(0, Math.min(100, score)));

  let patternClass: 'Healthy' | 'Distracted' | 'Compulsive_Addiction' = 'Healthy';
  if (finalScore >= 65) {
    patternClass = 'Compulsive_Addiction';
  } else if (finalScore >= 35) {
    patternClass = 'Distracted';
  }

  return {
    burnoutContributionScore: finalScore,
    usagePatternClass: patternClass,
    recommendedScreenCapHours: Math.round(Math.max(1.5, Math.min(8.0, cap)) * 10) / 10,
  };
}

// ---------------------------------------------------------------------
// 3. Cognitive Exhaustion & Focus Model
// ---------------------------------------------------------------------
export interface CognitiveFocusMLPrediction {
  predictedFocusScore: number;
  mentalExhaustionProbability: number;
  recommendedBreakIntervalMinutes: number;
}

export function predictCognitiveExhaustion(
  workHours: number,
  studyHours: number,
  exerciseMin: number,
  breakCount: number,
  sleepLastNight: number
): CognitiveFocusMLPrediction {
  const totalProd = workHours + studyHours;
  const breakRatio = totalProd > 0 ? (breakCount * 10) / (totalProd * 60) : 0.2;

  const inputs = [
    workHours,
    studyHours,
    exerciseMin,
    breakCount,
    breakRatio,
    sleepLastNight,
  ];

  const focus = evaluateLinearModel(
    activityModel.focus_model.coefficients,
    activityModel.focus_model.intercept,
    inputs
  );

  const exhaustion = evaluateLinearModel(
    activityModel.exhaustion_model.coefficients,
    activityModel.exhaustion_model.intercept,
    inputs
  );

  const finalFocus = Math.round(Math.max(10, Math.min(100, focus)));
  const finalExhaustion = Math.round(Math.max(0, Math.min(1.0, exhaustion)) * 100) / 100;
  const recommendedInterval = Math.round(Math.max(25, Math.min(60, 52 - totalProd * 2 + breakCount * 2)));

  return {
    predictedFocusScore: finalFocus,
    mentalExhaustionProbability: finalExhaustion,
    recommendedBreakIntervalMinutes: recommendedInterval,
  };
}

// ---------------------------------------------------------------------
// 4. Emotional Stability & Stress Intervention Model
// ---------------------------------------------------------------------
export interface EmotionStressMLPrediction {
  stabilityIndex: number;
  stressCategory: 'Calm' | 'Mild_Stress' | 'High_Anxiety' | 'Burnout_Distress';
  interventionUrgency: 'None' | 'Low' | 'Immediate_Breathing_Required';
}

const EMOTION_MAP: Record<string, number> = {
  Happy: 0,
  Sad: 1,
  Angry: 2,
  Neutral: 3,
  Surprised: 4,
  Anxious: 5,
};

export function predictEmotionStress(
  dominantEmotion: string,
  selfReportedStress: number, // 1 to 10
  valence: number, // -1.0 to 1.0
  arousal: number, // 0.0 to 1.0
  timeOfDay: number // 0 to 23.9
): EmotionStressMLPrediction {
  const code = EMOTION_MAP[dominantEmotion] ?? 3;
  const inputs = [code, selfReportedStress, valence, arousal, timeOfDay];

  const stability = evaluateLinearModel(
    emotionModel.stability_model.coefficients,
    emotionModel.stability_model.intercept,
    inputs
  );

  let category: 'Calm' | 'Mild_Stress' | 'High_Anxiety' | 'Burnout_Distress' = 'Calm';
  if (selfReportedStress >= 8) category = 'Burnout_Distress';
  else if (selfReportedStress >= 6) category = 'High_Anxiety';
  else if (selfReportedStress >= 4) category = 'Mild_Stress';

  let urgency: 'None' | 'Low' | 'Immediate_Breathing_Required' = 'None';
  if (selfReportedStress >= 8) urgency = 'Immediate_Breathing_Required';
  else if (selfReportedStress >= 6) urgency = 'Low';

  return {
    stabilityIndex: Math.round(Math.max(10, Math.min(100, stability))),
    stressCategory: category,
    interventionUrgency: urgency,
  };
}

// ---------------------------------------------------------------------
// 5. Master Burnout Forecaster
// ---------------------------------------------------------------------
export interface MasterBurnoutMLPrediction {
  masterBurnoutScore: number;
  future7DayScore: number;
  riskLevel: 'low' | 'moderate' | 'high' | 'critical';
  primaryDriver: 'Sleep' | 'Phone_Overuse' | 'Work_Overload' | 'Emotional_Distress';
}

export function predictMasterBurnout(
  avgSleepQual7d: number,
  avgSleepDur7d: number,
  avgScreenTime7d: number,
  avgPickups7d: number,
  avgWorkHours7d: number,
  avgExercise7d: number,
  negEmotionRatio7d: number,
  emotionStability7d: number
): MasterBurnoutMLPrediction {
  const inputs = [
    avgSleepQual7d,
    avgSleepDur7d,
    avgScreenTime7d,
    avgPickups7d,
    avgWorkHours7d,
    avgExercise7d,
    negEmotionRatio7d,
    emotionStability7d,
  ];

  const masterScore = evaluateLinearModel(
    masterModel.master_score_model.coefficients,
    masterModel.master_score_model.intercept,
    inputs
  );

  const futureScore = evaluateLinearModel(
    masterModel.future_7d_model.coefficients,
    masterModel.future_7d_model.intercept,
    inputs
  );

  const finalScore = Math.round(Math.max(0, Math.min(100, masterScore)));
  const finalFuture = Math.round(Math.max(0, Math.min(100, futureScore)));

  let riskLevel: 'low' | 'moderate' | 'high' | 'critical' = 'low';
  if (finalScore >= 75) riskLevel = 'critical';
  else if (finalScore >= 50) riskLevel = 'high';
  else if (finalScore >= 25) riskLevel = 'moderate';

  // Drivers breakdown
  const sleepContrib = (100 - avgSleepQual7d) * 0.25;
  const phoneContrib = avgScreenTime7d * 3.8;
  const workContrib = avgWorkHours7d * 3.2;
  const emotionContrib = negEmotionRatio7d * 22;

  const maxContrib = Math.max(sleepContrib, phoneContrib, workContrib, emotionContrib);
  let primaryDriver: 'Sleep' | 'Phone_Overuse' | 'Work_Overload' | 'Emotional_Distress' = 'Sleep';
  if (maxContrib === phoneContrib) primaryDriver = 'Phone_Overuse';
  else if (maxContrib === workContrib) primaryDriver = 'Work_Overload';
  else if (maxContrib === emotionContrib) primaryDriver = 'Emotional_Distress';

  return {
    masterBurnoutScore: finalScore,
    future7DayScore: finalFuture,
    riskLevel,
    primaryDriver,
  };
}

File: mobile/android/app/src/main/java/com/burnoutai/mobile/MLEngine.kt
package com.burnoutai.mobile

import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * Native Kotlin ML Engine for Burnout Detection and Wellness Predictions.
 * Runs lightweight linear regression & classification predictions on-device.
 */
object MLEngine {

    data class SleepMLPrediction(
        val predictedQualityScore: Int,
        val fatigueIndex: Int,
        val circadianDisruptionLevel: String
    )

    data class DigitalAddictionMLPrediction(
        val burnoutContributionScore: Int,
        val usagePatternClass: String,
        val recommendedScreenCapHours: Double
    )

    data class CognitiveFocusMLPrediction(
        val predictedFocusScore: Int,
        val mentalExhaustionProbability: Double,
        val recommendedBreakIntervalMinutes: Int
    )

    data class EmotionStressMLPrediction(
        val stabilityIndex: Int,
        val stressCategory: String,
        val interventionUrgency: String
    )

    data class MasterBurnoutMLPrediction(
        val masterBurnoutScore: Int,
        val future7DayScore: Int,
        val riskLevel: String,
        val primaryDriver: String
    )

    private fun evaluateLinearModel(coefficients: DoubleArray, intercept: Double, features: DoubleArray): Double {
        var result = intercept
        val count = min(coefficients.size, features.size)
        for (i in 0 until count) {
            result += coefficients[i] * features[i]
        }
        return result
    }

    /**
     * 1. Sleep Quality & Recovery Predictor (Kotlin Native)
     */
    fun predictSleepQuality(
        bedtimeHour: Double,
        wakeTimeHour: Double,
        durationHours: Double,
        interruptions: Int,
        prevScreenTime: Double,
        prevLateNight: Boolean,
        prevExercise: Double,
        prevStress: Double
    ): SleepMLPrediction {
        val coefsQuality = doubleArrayOf(-1.25, 0.85, 4.20, -5.50, -1.80, -10.0, 0.12, -2.80)
        val interceptQuality = 82.5

        val coefsFatigue = doubleArrayOf(0.95, -0.65, -3.10, 4.20, 1.45, 8.50, -0.09, 2.30)
        val interceptFatigue = 45.0

        val inputs = doubleArrayOf(
            bedtimeHour,
            wakeTimeHour,
            durationHours,
            interruptions.toDouble(),
            prevScreenTime,
            if (prevLateNight) 1.0 else 0.0,
            prevExercise,
            prevStress
        )

        val rawQuality = evaluateLinearModel(coefsQuality, interceptQuality, inputs)
        val rawFatigue = evaluateLinearModel(coefsFatigue, interceptFatigue, inputs)

        val quality = max(10, min(100, rawQuality.roundToInt()))
        val fatigue = max(0, min(100, rawFatigue.roundToInt()))

        var disruption = "Low"
        if ((bedtimeHour in 1.0..5.0) || prevLateNight) {
            disruption = if (interruptions > 2) "High" else "Moderate"
        }

        return SleepMLPrediction(quality, fatigue, disruption)
    }

    /**
     * 2. Digital Overuse & Addiction Classifier (Kotlin Native)
     */
    fun predictDigitalAddiction(
        totalScreen: Double,
        socialMedia: Double,
        entertainment: Double,
        productive: Double,
        pickups: Int,
        lateNight: Boolean,
        avgSessionMin: Double
    ): DigitalAddictionMLPrediction {
        val coefsScore = doubleArrayOf(4.2, 2.8, 1.9, -1.8, 0.14, 14.0, 0.25)
        val interceptScore = 8.5

        val coefsCap = doubleArrayOf(-0.35, -0.15, -0.10, 0.12, -0.01, -0.80, -0.02)
        val interceptCap = 6.2

        val inputs = doubleArrayOf(
            totalScreen,
            socialMedia,
            entertainment,
            productive,
            pickups.toDouble(),
            if (lateNight) 1.0 else 0.0,
            avgSessionMin
        )

        val rawScore = evaluateLinearModel(coefsScore, interceptScore, inputs)
        val rawCap = evaluateLinearModel(coefsCap, interceptCap, inputs)

        val finalScore = max(0, min(100, rawScore.roundToInt()))
        val patternClass = when {
            finalScore >= 65 -> "Compulsive_Addiction"
            finalScore >= 35 -> "Distracted"
            else -> "Healthy"
        }

        val roundedCap = (max(1.5, min(8.0, rawCap)) * 10).roundToInt() / 10.0

        return DigitalAddictionMLPrediction(finalScore, patternClass, roundedCap)
    }

    /**
     * 3. Cognitive Exhaustion & Focus Model (Kotlin Native)
     */
    fun predictCognitiveExhaustion(
        workHours: Double,
        studyHours: Double,
        exerciseMin: Double,
        breakCount: Int,
        sleepLastNight: Double
    ): CognitiveFocusMLPrediction {
        val totalProd = workHours + studyHours
        val breakRatio = if (totalProd > 0) (breakCount * 10.0) / (totalProd * 60.0) else 0.2

        val coefsFocus = doubleArrayOf(-4.2, -3.5, 0.15, 2.8, 12.0, 4.8)
        val interceptFocus = 65.0

        val coefsExhaust = doubleArrayOf(0.08, 0.06, -0.003, -0.04, -0.15, -0.07)
        val interceptExhaust = 0.35

        val inputs = doubleArrayOf(workHours, studyHours, exerciseMin, breakCount.toDouble(), breakRatio, sleepLastNight)

        val rawFocus = evaluateLinearModel(coefsFocus, interceptFocus, inputs)
        val rawExhaust = evaluateLinearModel(coefsExhaust, interceptExhaust, inputs)

        val finalFocus = max(10, min(100, rawFocus.roundToInt()))
        val finalExhaust = (max(0.0, min(1.0, rawExhaust)) * 100).roundToInt() / 100.0
        val recInterval = max(25, min(60, (52 - totalProd * 2 + breakCount * 2).toInt()))

        return CognitiveFocusMLPrediction(finalFocus, finalExhaust, recInterval)
    }

    /**
     * 4. Emotional Stability & Stress Intervention Model (Kotlin Native)
     */
    fun predictEmotionStress(
        dominantEmotion: String,
        selfReportedStress: Int,
        valence: Double,
        arousal: Double,
        timeOfDay: Double
    ): EmotionStressMLPrediction {
        val emotionCode = when (dominantEmotion.lowercase()) {
            "happy" -> 0.0
            "sad" -> 1.0
            "angry" -> 2.0
            "neutral" -> 3.0
            "surprised" -> 4.0
            "anxious" -> 5.0
            else -> 3.0
        }

        val coefsStab = doubleArrayOf(-1.2, -3.8, 18.0, -14.0, -0.15)
        val interceptStab = 78.0

        val inputs = doubleArrayOf(emotionCode, selfReportedStress.toDouble(), valence, arousal, timeOfDay)
        val rawStab = evaluateLinearModel(coefsStab, interceptStab, inputs)

        val finalStab = max(10, min(100, rawStab.roundToInt()))
        val category = when {
            selfReportedStress >= 8 -> "Burnout_Distress"
            selfReportedStress >= 6 -> "High_Anxiety"
            selfReportedStress >= 4 -> "Mild_Stress"
            else -> "Calm"
        }

        val urgency = when {
            selfReportedStress >= 8 -> "Immediate_Breathing_Required"
            selfReportedStress >= 6 -> "Low"
            else -> "None"
        }

        return EmotionStressMLPrediction(finalStab, category, urgency)
    }

    /**
     * 5. Master Burnout Forecaster (Kotlin Native)
     */
    fun predictMasterBurnout(
        avgSleepQual7d: Double,
        avgSleepDur7d: Double,
        avgScreenTime7d: Double,
        avgPickups7d: Double,
        avgWorkHours7d: Double,
        avgExercise7d: Double,
        negEmotionRatio7d: Double,
        emotionStability7d: Double
    ): MasterBurnoutMLPrediction {
        val coefsScore = doubleArrayOf(-0.25, -2.1, 3.8, 0.08, 3.2, -0.18, 22.0, -0.14)
        val interceptScore = 42.0

        val coefsFuture = doubleArrayOf(-0.27, -2.3, 4.0, 0.09, 3.4, -0.19, 23.5, -0.15)
        val interceptFuture = 44.0

        val inputs = doubleArrayOf(
            avgSleepQual7d,
            avgSleepDur7d,
            avgScreenTime7d,
            avgPickups7d,
            avgWorkHours7d,
            avgExercise7d,
            negEmotionRatio7d,
            emotionStability7d
        )

        val rawScore = evaluateLinearModel(coefsScore, interceptScore, inputs)
        val rawFuture = evaluateLinearModel(coefsFuture, interceptFuture, inputs)

        val finalScore = max(0, min(100, rawScore.roundToInt()))
        val finalFuture = max(0, min(100, rawFuture.roundToInt()))

        val riskLevel = when {
            finalScore >= 75 -> "critical"
            finalScore >= 50 -> "high"
            finalScore >= 25 -> "moderate"
            else -> "low"
        }

        val sleepContrib = (100 - avgSleepQual7d) * 0.25
        val phoneContrib = avgScreenTime7d * 3.8
        val workContrib = avgWorkHours7d * 3.2
        val emotionContrib = negEmotionRatio7d * 22.0

        val maxContrib = maxOf(sleepContrib, phoneContrib, workContrib, emotionContrib)
        val primaryDriver = when (maxContrib) {
            phoneContrib -> "Phone_Overuse"
            workContrib -> "Work_Overload"
            emotionContrib -> "Emotional_Distress"
            else -> "Sleep"
        }

        return MasterBurnoutMLPrediction(finalScore, finalFuture, riskLevel, primaryDriver)
    }
}
